lfs/api.go

```diff
@@ -0,0 +1,93 @@
+package lfs
+
+import (
+	"io"
+	"sync"
+)
+
+// An abstract interface providing state & resource management for a specific Endpoint across
+// potentially multiple requests
+type ApiContext interface {
+	// Get the endpoint this context was constructed from
+	Endpoint() Endpoint
+	// Close the context & any resources it's using
+	Close() error
+
+	// Download a single object, return reader for data, size and any error
+	// Essentially the same as calling DownloadCheck() then DownloadObject()
+	Download(oid string) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an object is available for download and return an object resource if so
+	DownloadCheck(oid string) (*ObjectResource, *WrappedError)
+	// Download a single object from an already identified resource from DownloadCheck(), return reader for data, size and any error
+	DownloadObject(obj *ObjectResource) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an upload would be accepted for an object and return the resource to use if so
+	UploadCheck(oid string, sz int64) (*ObjectResource, *WrappedError)
+	// Perform the actual upload of an object having identified it will be accepted and the resource to use
+	UploadObject(o *ObjectResource, reader io.Reader) *WrappedError
+	// Perform a batch request for a number of objects to determine what can be uploaded/downloaded
+	Batch(objects []*ObjectResource) ([]*ObjectResource, *WrappedError)
```

Comment:
OK - I was just working with what was there, happy for it to be 100% batch oriented too. In theory yeah that would eliminate the need for cached contexts within the usage model of the command line, although it does no harm for them to be there really; the SSH transport will always need a discrete setup & tear-down step anyway, so whether it's potentially around multiple ops or just one is neither here nor there in terms of added complexity.